sim/dex-data.ts

```diff
@@ -90,105 +92,82 @@ export class BasicEffect {
-	/**
-	 * Whether or not the effect is ignored by Baton Pass.
-	 */
+	/** Whether or not the effect is ignored by Baton Pass. */
 	noCopy: boolean;
-	/**
-	 * Whether or not the effect affects fainted Pokemon.
-	 */
+	/** Whether or not the effect affects fainted Pokemon. */
 	affectsFainted: boolean;
-	/**
-	 * The status that the effect may cause.
-	 */
-	status?: string;
-	/**
-	 * The weather that the effect may cause.
-	 */
-	weather?: undefined;
-	/**
-	 * HP that the effect may drain.
-	 */
-	drain?: number[];
+	/** The status that the effect may cause. */
+	status?: string; // TODO: Status
+	/** The weather that the effect may cause. */
+	weather?: undefined; // TODO: Weather
+	/** HP that the effect may drain. */
+	drain?: number[]; // TODO [number, number] or Fraction
 	flags: AnyObject;
 	sourceEffect: string;
 
-	constructor(data: AnyObject, moreData: AnyObject | null = null, moreData2: AnyObject | null = null) {
-		this.id = '';
-		this.name = '';
-		this.fullname = '';
-		this.effectType = 'Effect';
-		this.exists = true;
-		this.num = 0;
-		this.gen = 0;
-		this.isUnreleased = false;
-		this.shortDesc = '';
-		this.desc = '';
-		this.isNonstandard = false;
-		this.duration = this.duration;
-		// @ts-ignore
-		this.noCopy = !!this.noCopy;
-		// @ts-ignore
-		this.affectsFainted = !!this.affectsFainted;
-		this.status = this.status || undefined;
-		this.weather = this.weather || undefined;
-		this.drain = this.drain || undefined;
-		// @ts-ignore
-		this.flags = this.flags || {};
-		// @ts-ignore
-		this.sourceEffect = this.sourceEffect || '';
-
-		Object.assign(this, data);
-		if (moreData) Object.assign(this, moreData);
-		if (moreData2) Object.assign(this, moreData2);
-		this.name = Tools.getString(this.name).trim();
-		this.fullname = Tools.getString(this.fullname) || this.name;
-		if (!this.id) this.id = toId(this.name); // Hidden Power hack
-		// @ts-ignore
-		this.effectType = Tools.getString(this.effectType) || "Effect";
-		this.exists = !!(this.exists && this.id);
+	constructor(data: AnyObject) {
+		this.name = Tools.getString(data.name).trim();
+		this.id = data.id || toId(this.name); // Hidden Power hack
+		this.fullname = Tools.getString(data.fullname) || this.name;
+		this.effectType = Tools.getString(data.effectType) as EffectTypes || 'Effect';
+		this.exists = !!((data.hasOwnProperty('exists') ? data.exists : true) && this.id);
+		this.num = data.num || 0;
+		this.gen = data.gen || 0;
+		this.isUnreleased = data.isUnreleased || false;
+		this.shortDesc = data.shortDesc || '';
+		this.desc = data.desc || '';
+		this.isNonstandard = data.isNonStandard || false;
+		this.duration = data.duration;
+		this.noCopy = !!data.noCopy;
+		this.affectsFainted = !!data.affectsFainted;
+		this.status = data.status || undefined;
+		this.weather = data.weather || undefined;
+		this.drain = data.drain || undefined;
+		this.flags = data.flags || {};
+		this.sourceEffect = data.sourceEffect || '';
+		Object.assign(this, data); // TODO
```

Comment:
I'd like to remove this `Object.assign(...)` (and everywhere else in this file except `combine`)

Reply:
I think it should be retained – it's useful to support people adding their own attributes if they want.

Reply:
Alright, had to sleep on this.

I had trouble with `Object.assign` because it makes it difficult to understand/reason about which attributes are actually expected to be present. However, I think I'm just getting confused due to the interface merging going on, because the type of `BasicEffect` here also includes the fields from `BasicEffect` and `EffectData` and `EventMethods` etc as declared in `dev-tools/globals.ts`. 

Having the split type definitions (some bits in `globals.ts` and some in the file itself) is what my real issue is, not `Object.assign`. `Object.assign` is definitely required, because there's no way I'm going to be able to enumerate all of the fields (like the 148 possible listeners etc) in the constructor. 

**How would you feel about changing `BasicEffect` in this file to have a redundant `implements EffectData` in the class implementation**, as a hint to the reader about what is going on (which causes them to look for where `EffectData` is defined etc)? It should not effect behavior at all, but would have been very helpful to me as an outside observer to understand what is going on.

Reply:
I decided to go ahead with this everywhere. I don't know why, but it seems to have made Typescript better at typechecking (it noticed a `string` that should have been `string[]` only after I added an `implements`)

Reply:
That seems fine by me.

Reply:
TypeScript currently has a naming conflict between e.g. the class `Ability` and the interface `Ability`. They should probably be merged at some point.

Reply:
Ack